src/Microsoft.AspNetCore.ResponseCaching/ResponseOutput/ResponseCacheStream.cs

```diff
@@ -12,37 +12,56 @@ internal class ResponseCacheStream : Stream
     {
         private readonly Stream _innerStream;
         private readonly long _maxBufferSize;
+        private readonly int _shardSize;
+        private WriteOnlyShardStream _writeOnlyStream;
+        private ReadOnlyShardStream _readOnlyStream;
 
-        public ResponseCacheStream(Stream innerStream, long maxBufferSize)
+        internal ResponseCacheStream(Stream innerStream, long maxBufferSize, int shardSize)
         {
             _innerStream = innerStream;
             _maxBufferSize = maxBufferSize;
+            _shardSize = shardSize;
+            _writeOnlyStream = new WriteOnlyShardStream(_shardSize);
         }
 
-        public MemoryStream BufferedStream { get; } = new MemoryStream();
-
-        public bool BufferingEnabled { get; set; } = true;
+        internal bool BufferingEnabled { get; private set; } = true;
 
         public override bool CanRead => _innerStream.CanRead;
 
         public override bool CanSeek => _innerStream.CanSeek;
 
         public override bool CanWrite => _innerStream.CanWrite;
 
-        public override long Length => _innerStream.Length;
+        public override long Length => _writeOnlyStream.Length;
 
         public override long Position
         {
             get { return _innerStream.Position; }
-            set { _innerStream.Position = value; }
+            set
+            {
+                DisableBuffering();
+                _innerStream.Position = value;
+            }
+        }
+
+        internal Stream GetShardStream()
+        {
+            if (_readOnlyStream == null)
```

Comment:
I don't expect it to be called multiple times. But if we did, this is actually wrong since we want to recreate a new ReadOnlyShardStream since reads are stateful.